src/server/pfs/s3/multipart.go

```diff
@@ -0,0 +1,412 @@
+package s3
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/gorilla/mux"
+	"github.com/pachyderm/pachyderm/src/client"
+	pfsClient "github.com/pachyderm/pachyderm/src/client/pfs"
+	pfsServer "github.com/pachyderm/pachyderm/src/server/pfs"
+	"github.com/pachyderm/pachyderm/src/server/pkg/errutil"
+	"github.com/pachyderm/pachyderm/src/server/pkg/uuid"
+	"github.com/pachyderm/s2"
+)
+
+var multipartChunkPathMatcher = regexp.MustCompile(`([^/]+)/([^/]+)/(.+)/([^/]+)/(\d+)`)
+var multipartKeepPathMatcher = regexp.MustCompile(`([^/]+)/([^/]+)/(.+)/([^/]+)/\.keep`)
+
+func multipartChunkArgs(path string) (repo string, branch string, key string, uploadID string, partNumber int, err error) {
+	match := multipartChunkPathMatcher.FindStringSubmatch(path)
+
+	if len(match) == 0 {
+		err = errors.New("invalid file path found in multipath bucket")
+		return
+	}
+
+	repo = match[1]
+	branch = match[2]
+	key = match[3]
+	uploadID = match[4]
+	partNumber, err = strconv.Atoi(match[5])
+	if err != nil {
+		err = fmt.Errorf("invalid file path found in multipath bucket: %s", err)
+		return
+	}
+	return
+}
+
+func multipartKeepArgs(path string) (repo string, branch string, key string, uploadID string, err error) {
+	match := multipartKeepPathMatcher.FindStringSubmatch(path)
+
+	if len(match) == 0 {
+		err = errors.New("invalid file path found in multipath bucket")
+		return
+	}
+
+	repo = match[1]
+	branch = match[2]
+	key = match[3]
+	uploadID = match[4]
+	return
+}
+
+func parentDirPath(repo, branch, key, uploadID string) string {
+	return fmt.Sprintf("%s/%s/%s/%s", repo, branch, key, uploadID)
```

Comment:
For path construction I'd recommend using the `path` package, for one it's a tad shorter, this would be written as `path.Join(repo, branch, key, uploadID)` but more importantly without it I'm always worried I'll accidentally wind up with a double slash or a missing slash, whereas with `path.Join` you don't have to think about it.